source/dotnet/Library/AdaptiveCards/AdaptiveBaseElement.cs

```diff
@@ -0,0 +1,99 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Xml.Serialization;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace AdaptiveCards
+{
+    public class AdaptiveInternalID
+    {
+        public static uint Invalid = 0;
+        private static uint CurrentInternalID = 1;
+        private uint ID { get; }
+
+        public AdaptiveInternalID()
+        {
+            ID = Invalid;
+        }
+
+        private AdaptiveInternalID(uint id)
+        {
+            ID = id;
+        }
+
+        public static AdaptiveInternalID Current()
+        {
+            return new AdaptiveInternalID(CurrentInternalID);
+        }
+
+        public static AdaptiveInternalID Next()
+        {
+            CurrentInternalID++;
+
+            // handle overflow case
+            if (CurrentInternalID == AdaptiveInternalID.Invalid)
```

Comment:
Its not very intuitive to have a check against an "invalid" value and then make it "valid" by incrementing what is supposed to be invalid :( ; Seems scary. It just so happens though that the "invalid" number is just simply 0 so this will be less scary in practice and more normal. 
i would consider renaming Invalid -  to just 'defaultId' or even hard code the check against 0 directly.

Reply:
a few things of note here -- Carlos adapted this directly from C++. The overflow case is where a single instance of AdaptiveCards has processed >`UINT_MAX` elements (which would take... quite some time). This code isn't correct as written as I assume that doing:

```
uint i = UINT_MAX; // or whatever it is in C#
i++;
```

will throw an exception. I'll adapt it to be more correct. :)

---
In reply to: [278316341](https://github.com/Microsoft/AdaptiveCards/pull/2560#discussion_r278316341) [](ancestors = 278316341)